### PR TITLE
feature/tas yml segregation

### PR DESCRIPTION
**Pull request by VikrantKS**
# Issue Link
https://github.com/LambdaTest/test-at-scale/issues/224
# Description

This PR will segregate tas yml parsing from discovery 

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Tested with cloud runner for bit bucket, gitlab and github
- [x] Tested with synapse for  bit bucket, gitlab and github

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pkg/core/models.go [hunk 1]
```diff
@@ -456,3 +454,29 @@ type TestExecutionArgs struct {
...
+	CommitID    string    `json:"commitID"`
+	Event       EventType `json:"eventType"`
+	RepoSlug    string    `json:"repoSlug"`
+	TasFileName string    `json:"tasFilePath"`
+	LicenseTier Tier      `json:"license_tier"`
+	OrgID       string    `json:"orgID"`
+	BuildID     string    `json:"buildID"`
+}
+
+// TASConfigDownloaderOutput repersent output return by by tasconfig downloader
```
Comment (saurabh-prakash) [inline pkg/core/models.go:470]:
```suggestion
// TASConfigDownloaderOutput repersent output return by tasconfig downloader
```

File: pkg/core/models.go [hunk 2]
```diff
@@ -456,3 +454,29 @@ type TestExecutionArgs struct {
...
+type TASConfigDownloaderOutput struct {
+	Version   int         `json:"version"`
+	TasConfig interface{} `json:"tasConfig"`
```
Comment (saurabh-prakash) [inline pkg/core/models.go:473]:
```suggestion
	TASConfig interface{} `json:"tasConfig"`
```

Comment (saurabh-prakash) [reply pkg/core/models.go:473]: Won't `TASConfig` have a version in itself?

File: pkg/gitmanager/setup.go [hunk 3]
```diff
@@ -174,3 +166,53 @@ func (gm *gitManager) initGit(ctx context.Context, payload *core.Payload, oauth
 	}
 	return nil
 }
+
+func (gm *gitManager) makeDownloadRequest(ctx context.Context, downloadURL string,
```
Comment (saurabh-prakash) [inline pkg/gitmanager/setup.go:170]: Please use `requestutils` package instead of defining this new function

File: pkg/synapse/synapse.go [hunk 4]
```diff
@@ -376,3 +383,42 @@ func (s *synapse) messageWriter(conn *websocket.Conn) {
...
+	var parsingReqMsg *core.YMLParsingRequestMessage
+	var writeMsg core.Message
+	defer s.writeMessageToBuffer(&writeMsg)
+	if err := json.Unmarshal(message.Content, &parsingReqMsg); err != nil {
+		s.logger.Errorf("error in unmarshaling message for yml parsing request, error %v ", err)
+
+		writeMsg = CreateYMlParsingResultMessage(core.YMLParsingResultMessage{
+			OrgID:     parsingReqMsg.OrgID,
+			BuildID:   parsingReqMsg.BuildID,
+			YMLOutput: core.TASConfigDownloaderOutput{},
```
Comment (saurabh-prakash) [inline pkg/synapse/synapse.go:397]: 0 values need not be explicitly set.

File: pkg/synapse/synapse.go [hunk 5]
```diff
@@ -376,3 +383,42 @@ func (s *synapse) messageWriter(conn *websocket.Conn) {
...
+		parsingReqMsg.CommitID,
+		parsingReqMsg.RepoSlug, parsingReqMsg.TasFileName, oauth,
+		parsingReqMsg.Event, parsingReqMsg.LicenseTier)
+	if err != nil {
+		s.logger.Errorf("error occurred while fetching tas config file for buildID %s orgID %s, error %v",
+			parsingReqMsg.BuildID, parsingReqMsg.OrgID, err)
+		writeMsg = CreateYMlParsingResultMessage(core.YMLParsingResultMessage{
+			OrgID:     parsingReqMsg.OrgID,
+			BuildID:   parsingReqMsg.BuildID,
+			YMLOutput: core.TASConfigDownloaderOutput{},
```
Comment (saurabh-prakash) [inline pkg/synapse/synapse.go:414]: 0 values need not be explicitly set.

File: pkg/synapse/utils.go [hunk 6]
```diff
@@ -75,3 +75,16 @@ func GetResources(tierOpts core.Tier) core.Specs {
 	}
 	return core.Specs{CPU: 0, RAM: 0}
 }
+
+// CreateYMlParsingResultMessage creates message for YML parsing result
+func CreateYMlParsingResultMessage(ymlParsingOutput core.YMLParsingResultMessage) core.Message {
```
Comment (saurabh-prakash) [inline pkg/synapse/utils.go:80]: this is used within synapse package, therefore no need to export it.

File: pkg/tasconfigdownloader/setup.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+func New(logger lumber.Logger) *TASConfigDownloader {
+	return &TASConfigDownloader{
+		logger:           logger,
+		gitmanager:       gitmanager.NewGitManager(logger, nil),
+		tasconfigmanager: tasconfigmanager.NewTASConfigManager(logger),
+	}
+}
+
+func (t *TASConfigDownloader) GetTasConfig(ctx context.Context, gitProvider, commitID, repoSlug,
```
Comment (saurabh-prakash) [inline pkg/tasconfigdownloader/setup.go:27]:
```suggestion
func (t *TASConfigDownloader) GetTASConfig(ctx context.Context, gitProvider, commitID, repoSlug,
```

File: pkg/gitmanager/setup.go [hunk 8]
```diff
@@ -174,3 +166,53 @@ func (gm *gitManager) initGit(ctx context.Context, payload *core.Payload, oauth
...
+	downloadURL, err := urlmanager.GetFileDownloadURL(gitProvider, commitID, repoSlug, filePath)
+	if err != nil {
+		return "", err
+	}
+	resp, err := gm.makeDownloadRequest(ctx, downloadURL, oauth)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+	path := utils.GenerateUUID() + ".yml"
```
Comment (saurabh-prakash) [inline pkg/gitmanager/setup.go:204]: why a random path and not something like `<commit_id>/.tas.yml` in os.TempDir() folder? It'll cache the file for subsequent requests too. Also, use sync.Once to do this only once if file doesn't exist at the above path (See usage of `c.once.Do` in cachemanager).

File: pkg/utils/utils.go [hunk 9]
```diff
@@ -269,3 +270,24 @@ func GetArgs(command string, frameWork string, frameworkVersion int,
...
+	path, err := GetConfigFileName(path)
+	if err != nil {
+		return "", err
+	}
+	filePath := fmt.Sprintf("%s/%s", global.RepoDir, path)
+	return filePath, nil
+}
+
+// GenerateUUID generates uuid v4
+func GenerateUUID() string {
```
Comment (saurabh-prakash) [inline pkg/utils/utils.go:285]: this function may not be required.

File: pkg/utils/utils.go [hunk 10]
```diff
@@ -269,3 +270,24 @@ func GetArgs(command string, frameWork string, frameworkVersion int,
 
 	return args
 }
+
+// GetTasFilePath returns tas file path
+func GetTasFilePath(path string) (string, error) {
```
Comment (saurabh-prakash) [inline pkg/utils/utils.go:275]:
```suggestion
func GetTASFilePath(path string) (string, error) {
```

File: pkg/core/models.go (hunk 2 above)
Comment (VikrantKS) [reply pkg/core/models.go:473]: Both way is fine , in current approach , user can just check the version and parse YML accordingly
In second version , he has to parse YML twice, first to check version and then to get exact TAS config
I think first approach is simpler but open to change

File: pkg/gitmanager/setup.go (hunk 3 above)
Comment (VikrantKS) [reply pkg/gitmanager/setup.go:170]: This is refactor code, I have taken out common code from function `downloadFile`
Agree with your comment , will change it

File: pkg/gitmanager/setup.go (hunk 8 above)
Comment (VikrantKS) [reply pkg/gitmanager/setup.go:204]: I wanted to clean the path (delete the file) after the processing it, as there is no merit in keeping the file (also it occupies some space) , keeping commit-id in path may lead to race condition in that case

File: pkg/core/models.go (hunk 1 above)
Comment (VikrantKS) [reply pkg/core/models.go]: Done

File: pkg/synapse/synapse.go (hunk 4 above)
Comment (VikrantKS) [reply pkg/synapse/synapse.go]: Done

File: pkg/synapse/synapse.go (hunk 5 above)
Comment (VikrantKS) [reply pkg/synapse/synapse.go]: Done

File: pkg/tasconfigdownloader/setup.go (hunk 7 above)
Comment (VikrantKS) [reply pkg/tasconfigdownloader/setup.go]: Done

File: pkg/utils/utils.go (hunk 9 above)
Comment (VikrantKS) [reply pkg/utils/utils.go:287]: Need to discuss this , based on earlier comment

File: pkg/utils/utils.go (hunk 10 above)
Comment (VikrantKS) [reply pkg/utils/utils.go]: Done

**Review by saurabh-prakash — APPROVED** [review]

**Pull request merged**